src/ssl/support.cc

```diff
@@ -962,6 +963,10 @@ void
 Ssl::chainCertificatesToSSLContext(Security::ContextPointer &ctx, Security::ServerOptions &options)
 {
     assert(ctx);
+
+    if (Security::CertIsSelfSigned(*options.signingCa.cert))
+        return;
+
     // Add signing certificate to the certificates chain
     X509 *signingCert = options.signingCa.cert.get();
     if (SSL_CTX_add_extra_chain_cert(ctx.get(), signingCert)) {
```

Comment:
> SSL_CTX_add_extra_chain_cert() [documentation](https://www.openssl.org/docs/man1.1.1/man3/SSL_CTX_add_extra_chain_cert.html) says "When sending a certificate chain, extra chain certificates are sent in order following the end entity certificate."
> 
> To me, that documentation implies that the extra chain should not contain the signing certificate (i.e. the "end entity" certificate) itself. Why are we adding the signingCert here? If we do not have to add it, then we may be able to _remove_ this code _instead_ of adding more `if`s.

We have to add the signing certificate to the chain IF it is an intermediate certificate and not a root CA.
We are just using wrong OpenSSL function. But the result is (mostly) the same.
We should use the  SSL_CTX_add1_chain_cert or  SSL_CTX_add0_chain_cert instead of SSL_CTX_add_extra_chain_cert.

The SSL_CTX_add1_chain_cert/SSL_CTX_add0_chain_cert    functions does some checks to check if the certificates matches SSL and SSL_CTX configuration before add to the chain list.

But I think I am see the same checks for the certificates added to the chain using the SSL_CTX_add_extra_chain_cert before encoded into a TLS packet.

The SSL_CTX_add1_chain_cert/SSL_CTX_add0_chain_cert added to the OpenSSL API from the version openssl-1.1.0 releases so I believe the backward compatibility is the reason for all of these functions.

> 
> The same documentation says "If no chain is specified, the library will try to complete the chain from the available CA certificates in the trusted CA storage". I think that is OK because if the signing certificate is self-signed, then OpenSSL will not be able to "complete the chain" -- the single-certificate chain (containing just the signing certificate) is already completed.

> 
> N.B. I know that the loop below this line that iterates options.signingCa.chain and adds those certificates. I hope that, when we are done here, options.signingCa.chain will be empty when options.signingCa.cert is self-signed. Perhaps it already is empty in such cases? Thus, no special protection will be needed for that loop _if_ we can remove this SSL_CTX_add_extra_chain_cert() call.

Yes the options.signingCa.chain is empty if signingCa.cert is self-signed. Also it is empty if the signingCa.cert is signed with root CA certificate.

Reply:
> We have to add the signing certificate to the chain IF it is an intermediate certificate and not a root CA.

That answer does not compute for me for two reasons:

1. Even if the signing certificate is an intermediate certificate (and not a root CA) I do not understand why we are adding it. In other words, the "we have to add if ..." answer does not answer the "why add" question.
2. The signing certificate cannot be an intermediate certificate by definition of an intermediate certificate: AFAIK, an intermediate certificate is a certificate that signs the signing certificate (or another intermediate certificate) and that itself, is not a root certificate: signing certificate -- optional intermediate certificates -- optional root certificate.

If my definition of an intermediate certificate is wrong, please let me know. If it is correct, then let's come back my original question: _Why_ are we adding the signingCert to the chain here? In other words, if we remove that SSL_CTX_add_extra_chain_cert() call, then can you give an example of a specific use case that we will break?

P.S. I understand that we are using a legacy OpenSSL function that does fewer checks than some modern ones, and that is why we need more manual checks before we call that legacy function, but my question is why call that function (on this particular line, with the signing certificate) _at all_? The signing certificate we provide ought to be already known to OpenSSL because it is a _signing_ certificate that we have already provided in createSSLContext() or some such. What am I missing?

Reply:
Here is a quote from RFC [9216](https://www.rfc-editor.org/rfc/rfc9216#name-certificate-chains) that matches my understanding of what intermediate CA is, and that an intermediate certificate is _not_ the signing (a.k.a. end-entity) certificate:

> In most real-world examples, X.509 certificates are deployed with a chain of more than one X.509 certificate. In particular, there is typically a long-lived root CA that users' software knows about upon installation, and the end-entity certificate is issued by an intermediate CA, which is in turn issued by the root CA

There are similar quotes in many less authoritative sources although I cannot find a good authoritative definition of these terms (there are some deprecated ones in old RFC, but nothing modern/relevant that I can find).